### DISCUSSION TITLE
Zod schema serialize/deserialize

### DISCUSSION THREAD

**Discussion by valerii15298** [colinhacks/zod#2030] [Q&A] [answered]
How can I serialise/deserialize zod schema?

So that I can store it in database for example as a string and reuse later when needed.

**Comment by JacobWeisenburger** [comment]
Can you explain more about your use case? Perhaps a give a code example.

**Reply by dabdine** [reply]
I had been looking into this to see if it was a possibility and stumbled on this discussion, so let me describe mine:

We're using a SSR-based framework for serving our application, Remix. Remix has loaders and actions, which allow for reads and writes server-side, respectively.

In my particular use case, we have two separate sets of fields (one describing a user, the other describing an organization) that need to submit to a single action URL (the action in Remix). Implementation wise, that looks like two separate Zod objects that we then `.merge` to create a single object to validate the form.

In some cases we have the user portion, but not the organization portion. When we have the user portion, we don't want to ask/revalidate that information (long story short, it causes an issue with our form validator, remix-validated-form). Being able to make the user portion optional by checking the value of some variable (for example, look up the user in the database and if the user is defined, make the user portion optional) would be nice to define, but that happens server-side. Thus, being able to conditionally construct a zod definition server side, then serialize it to the client (client component) would be helpful.

However, there are workarounds. The easiest workaround is to simply compute this with the non-server side components by sending the user portion (in our case) to the client and performing the calculation there.

In our case, we wouldn't need a database to store the schema in (in fact, we use json schema for that because we have golang in our stack, but I admit Zod is much nicer to deal with).

**Reply by sp88011** [reply]
Another use case is when you host a form builder (and the eventual form that is built with it). Say you want to allow users who build forms on your site to apply different logic to how the submitted data is transformed. 

Storing everything (validation, error-messages, transformation) in a single zod object is elegant, but because it's user-generated it needs to be saved in a DB.

**Reply by valerii15298** [reply]
@sp88011 yeah, I have exactly the same use case. For now I just decided to store zod schema as raw code string and then use JavaScript `eval` function

**Reply by sp88011** [reply]
@valerii15298 I'm curious how you do that.. any chance you can share some snippets? Particularly, how do you turn the Zod object (that presumably was user-generated) into a string that you can later use `eval()` on?

**Reply by valerii15298** [reply]
```js
const schemaString = `
// you can put any helper function or code directly inside the string and use them in your schema

function getZodSchema({z, ctx}) {
  // use ctx for any dynamic data that your schema depends on
  return z.object({
    id: z.string()
  });
}
`;

const schema = Function(
  "...args",
  `${schemaString}; return getZodSchema(...args)`,
)({ z, ctx: {} });
```

**Comment by JacobWeisenburger** [comment]
I don't think there is a built in way to serialize/deserialize a schema. However, you could definitely write something that would serialize/deserialize a schema. This would be fairly easy for most basic types (string, number, bool), but it might become difficult for more complex objects, records, or arrays.

**Comment by StefanTerdell** [comment, marked as answer]
This has come up a few times in [json-schema-to-zod](https://github.com/StefanTerdell/json-schema-to-zod), which along with [its counterpart](https://github.com/StefanTerdell/zod-to-json-schema) probably is the closest you'll get to what you're looking for. There has been a couple of requests for a way to do essentially what you're describing here - turn a JSON schema into a Zod schema at runtime. My stance has always been that it doesn't make much sense to turn a perfectly good JSON schema into a Zod schema due to possibly getting weaker validation caused by translation errors or limitations. For instance, there is no (sane) way to deserialize a preprocess callback, and JSTZ even utilizes superRefine to bridge the gap of Zods missing equivalent of the `oneOf`-keyword! 🤪 Since type inference wouldn't help you at runtime and you can just use the JSON schema with AJV directly it just doesn't make that much sense to me.

Then again, the request keeps popping up, and as the ecosystem grows it may start to make more sense even with the limitations present. 

At some point I'd love to find enough time to split json-schema-to-zod into two modules: one to serialize an actual Zod schema and one to stringify it. In the meantime you could just eval the output though.

**Reply by shacthulu** [reply]
Would serialization/deserialization at runtime using a file of typescript definitions be a viable option in the current implementation? It's obviously a subset of functionality offered by Zod by nature, but feels like it could cover the large majority of practical use cases with a readable syntax and offer a ton  of convenience for existing type files, auto documentation utilities, etc.

**Reply by nabilfreeman** [reply]
For anybody getting here, consider not using zod, instead implementing https://github.com/ThomasAribart/json-schema-to-ts for types and another library for JSON schema validation

**Comment by Gregoor** [comment]
I started working on a serialization library for zod that also retains type information. It can be found overe here: https://github.com/commonbaseapp/zodex

It's still a WIP, and there are still some battles with TypeScript I'll have to fight (if there's some wizards reading this, reach out if you want to help), but I'm hoping to stabilize soonish!

**Reply by brettz9** [reply]
Since not everyone wanting a serialization format is looking to work with JSON, converting to JSON Schema might not be ideal for all uses anyhow (e.g., an IndexedDB schema might wish to indicate a date is being stored rather than a date encoded as a string). Zodex sounds like a great beginning—the likes of which I hope may become blessed so there are not competing formats striving to do the same thing.

**Reply by janzheng** [reply]
this has been incredibly useful! I've been storing zod schemas as strings and this really helps

**Comment by theluk** [comment]
Zod, IMO is not made for serializing. If you use it heavily with refinements or processing functions, there is no way to serialize it other than stringifying code. I would suggest to refrain from using this approach. 

I would suggest sharing your Zod schemas. 

I have the following.

- shared/models/MyZodObject.ts
- app/ComponentUsingMyZodObject.tsx
- server/CodeUsingMyZodObject.tsx

Both app and server tsconfigs and bundlers include the shared into their projects. 

if you do not have a mono repo, you might still try using linking for example. Another approach is to publish your models to a private npm.

**Reply by brettz9** [reply]
imo, refinements are sometimes a sign that Zod could have further built-in improvements (e.g., toStringTag checks), but if https://github.com/colinhacks/zod/issues/2515 were implemented, there'd at least be a way to serialize those refinements, however non-standardly, so that a deserializer could handle them.

There really is a need for storing schemas specific to JavaScript objects (and in a safe way beyond executable code), and Zod (combined with Zodex) already goes very far to this end.

**Reply by vinayakakv** [reply]
We are planning to use feature flags powered by zod validation in our application and provide an UI to edit them.

The approach you suggested feels the best since we are using a monorepo setup already.

If the refinements/transformations use the variables that are outside schema's scope, serialization becomes nearly impossible.

**Comment by domyalex** [comment]
@StefanTerdell ,

+1 vote to supporting deserialization =)

Our company is now starting to work with MCP Server (all the rage now) which relies on Zod **object** when we register a tool ([https://github.com/modelcontextprotocol/typescript-sdk ](https://github.com/modelcontextprotocol/typescript-sdk )).

Due to the way our repositories are setup, unfortunately we can only flow text files from where we define the tool (with the Zod object) to where we need it to be applied to the MCP Server.

We can use zod-to-schema to serialize things, but on the other side we'd need a clean way to deserialize it into a full Zod object; Ajv is not an option because we are not talking about runtime input validation, but server startup tool registration.

At this point, short of using `eval`, we'd need to do something akin to

`
function jsonToZodSchema(jsonSchema) {
  if (jsonSchema.type === 'object') {
    const properties = jsonSchema.properties;
    const zodProperties = {};

    for (const key in properties) {
      const property = properties[key];
      switch (property.type) {
        case 'string':
          let zodString = z.string();
          if (property.format === 'email') {
            zodString = zodString.email();
          }
          zodProperties[key] = zodString;
          break;
        case 'integer':
          let zodNumber = z.number().int();
`


which we'd really, REALLY like to avoid.

**Reply by brettz9** [reply]
Have you looked at [zodex](https://github.com/commonbaseapp/zodex)?

**Reply by domyalex** [reply]
Ohh, excellent, thank you!

**Comment by Hussseinkizz** [comment]
Hmm I have looked at this issue also looking for perfect 2 way conversion, serialize and deserialize, my use case I want something like trpc but for non monorepo setting, I want our frontend to reference types or schema used on backend when intergrating apis without us having to have our frontends and backend in one repo as we have multi tenancy on both frontend and backend that's it would be an overkill if we put in mono repo and would be hard to mantain things!
